exchange/currency.go

```diff
@@ -0,0 +1,222 @@
+						// Check if conversion rates were provided in request first.
+						if requestExt.Currency.Rates != nil {
+							// If they are set in request, convert the currencies.
+							var convns conversions
+							if err := json.Unmarshal(bytes.ToUpper(requestExt.Currency.Rates), &convns); err != nil {
+								bidderErrs = append(bidderErrs, errors.New(defaultCurrencyConversionError).Error())
+								continue
+							}
+
+							if convertedBid, err = convertCurrency(convns, requestCurrency, bidCurrency, bid.Price); err != nil {
+								// Error converting currency with request rates. Try with latest rates.
+								if convertedBid, successfulConversion = convertCurrencyWithCurrentRates(&bidderErrs, requestCurrency, bidCurrency, bid); successfulConversion == false {
+									continue
+								}
+							}
+						} else {
+							// Conversion rates are not set in request. Use latest rates.
+							if convertedBid, successfulConversion = convertCurrencyWithCurrentRates(&bidderErrs, requestCurrency, bidCurrency, bid); successfulConversion == false {
+								continue
+							}
+						}
+
+						// Rate conversion successful. Set new bid price.
+						bid.Price = convertedBid
+					}
+					// Rate conversion was successful or no conversion was needed. Set new currency.
+					currencyToUse = requestCurrency
+					newBids = append(newBids, bid)
+				}
+
+				if len(newBids) > 0 {
+					// Set new bids for seat
+					seatBid.Bid = newBids
+					newSeatBids = append(newSeatBids, seatBid)
+				}
+
+				for _, bidderErr := range bidderErrs {
+					// Set errors for this seat
+					adapterExtraCopy[bidderName].Errors = append(adapterExtraCopy[bidderName].Errors, fmt.Sprintf("%s", bidderErr))
+				}
+				*adapterExtra = adapterExtraCopy
+			}
+			// Set the updated list of seatBids
+			*seatBids = newSeatBids
+		}
+	}
+
+	return currencyToUse
+}
+
+func convertCurrencyWithCurrentRates(bidderErrs *[]interface{}, requestCurrency string, bidCurrency string, bid openrtb.Bid) (float64, bool) {
+	var convertedBid float64
+	var successfulConversion bool
+	conversionRates := fetchCurrencyConversionRates()
+
+	if conversionRates == nil {
+		// If converstion rates not available, return error and drop the bid.
+		*bidderErrs = append(*bidderErrs, errors.New(defaultCurrencyConversionError).Error())
+	} else {
+		// Convert the currencies
+		var currData currencyData
+		if err := json.Unmarshal(conversionRates, &currData); err != nil {
+			*bidderErrs = append(*bidderErrs, errors.New(defaultCurrencyConversionError).Error())
+		} else {
+			convertedBid, err = convertCurrency(currData.Conversions, requestCurrency, bidCurrency, bid.Price)
+			if err != nil {
+				// Error converting currencies. Drop the bid.
+				*bidderErrs = append(*bidderErrs, err.Error())
+			} else {
+				successfulConversion = true
+			}
+		}
+	}
+
+	return convertedBid, successfulConversion
+}
+
+func fetchCurrencyConversionRates() []byte {
+	var rates []byte
```

Comment:
would like to cache this so that we're not fetching the rates from this url every time. having the file cached every 24 hours would be optimal. any ideas on how to do this? also wondering if there would be a significant improvement in performance.

Reply:
Agreed. Even with cacheing, there's no reason to take this performance hit _during_ an auction.

Best way I can think of:

1. Fetch the conversions once on app startup (with a blocking call in `pbs_light`. Don't let the server even bind to the port to start answering requests until it returns).
2. Stash the conversion data in an [atomic Value](https://golang.org/pkg/sync/atomic/#Value)
3. Start a separate goroutine which refreshes the data periodically using a [Ticker](https://golang.org/pkg/time/#Ticker).

The atomic swap will impact auctions _much_ less than the external HTTP call. A 24 hour period sounds reasonable... but if you take this approach, I don't think there's much harm in doing it more frequently either.

Reply:
added this functionality.